starknet_py/proxy/proxy_check.py

```diff
@@ -38,48 +35,18 @@ class ArgentProxyCheck(ProxyCheck):
     async def implementation_address(
         self, address: Address, client: Client
     ) -> Optional[int]:
-        return await self.get_implementation(
-            address=address,
-            client=client,
-            get_class_func=client.get_class_hash_at,
-            regex_err_msg=r"(is not deployed)",
+        call = Call(
+            to_addr=address,
+            selector=get_selector_from_name("get_implementation"),
+            calldata=[],
         )
+        (implementation,) = await client.call_contract(call=call)
+        return implementation
 
     async def implementation_hash(
         self, address: Address, client: Client
     ) -> Optional[int]:
-        return await self.get_implementation(
-            address=address,
-            client=client,
-            get_class_func=client.get_class_by_hash,
-            regex_err_msg=r"(is not declared)",
-        )
-
-    @staticmethod
-    async def get_implementation(
-        address: Address, client: Client, get_class_func: Callable, regex_err_msg: str
-    ) -> Optional[int]:
-        call = ArgentProxyCheck._get_implementation_call(address=address)
-        err_msg = r"(Entry point 0x[0-9a-f]+ not found in contract)|" + regex_err_msg
-        try:
-            (implementation,) = await client.call_contract(call=call)
-            await get_class_func(implementation)
-        except ClientError as err:
-            if (
-                re.search(err_msg, err.message, re.IGNORECASE)
-                or err.code == RPC_INVALID_MESSAGE_SELECTOR_ERROR
-            ):
-                return None
-            raise err
-        return implementation
-
-    @staticmethod
-    def _get_implementation_call(address: Address) -> Call:
-        return Call(
-            to_addr=address,
-            selector=get_selector_from_name("get_implementation"),
-            calldata=[],
-        )
+        return await self.implementation_address(address, client)
```

Comment:
How do we distinguish if the returned value was address or class hash now?

Reply:
We don't.
Higher level will try to get class_hash of implementation by calling `ProxyCheck.implementation_hash` and then, if no success, will try to get address by calling `ProxyCheck.implementation_address`.

Reply:
Okay, makes sense. But if a user wants to define a custom proxy check where they distinguish between hash and address, that would still work right?

Reply:
Shouldn't we return `class_hash` in the `implementation_hash`? It is confusing when `implementation_hash` returns an address 😕

Reply:
**OpenZeppelinProxyCheck** has different implementations of `implementation_address` and `implementation_hash`

Reply:
> Shouldn't we return class_hash in the implementation_hash? It is confusing when implementation_hash returns an address 😕

The methods are just instructions for the resolver on how to retrieve the class_hash or address from the Proxy.

Reply:
I know, but I can't see the profit of making `implementation_hash` returning an address

Reply:
It just means that the methods have the same logic. If you mean that it does not make much sense for the `implementation_hash` method to return some *address*, you're right.
I will add one more function and call it from both `implementation_hash` and `implementation_address`.